templates/support/SecurityClient.swift.hbs

```diff
@@ -0,0 +1,361 @@
+}
+
+/** Delegates security requiements to the appropriate SecurityClient. */
+public class SecurityClientController: SecurityClient {
+    public func authenticationFailed(securityScheme: SecurityScheme, scopes: [String]?) async throws {
+        if let client = clients[securityScheme] {
+            try await client.authenticationFailed(securityScheme: securityScheme, scopes: scopes)
+        } else {
+            throw APIError.securitySchemeNotConfigured(securityScheme)
+        }
+    }
+
+    public func authorize(securityScheme: SecurityScheme, scopes: [String]?, request: URLRequest) async throws -> URLRequest {
+        if let client = clients[securityScheme] {
+            return try await client.authorize(securityScheme: securityScheme, scopes: scopes, request: request)
+        } else {
+            return request
+        }
+    }
+    
+    var clients: [SecurityScheme: SecurityClient] = [:]
+    
+    public func setClient(securityScheme: SecurityScheme, client: SecurityClient) {
+        self.clients[securityScheme] = client
+    }
+
+    public init(clients: [SecurityScheme : SecurityClient] = [:]) {
+        self.clients = clients
+    }
+}
+
+func createOAuthRequest(url: URL, params: [String: String]) -> URLRequest {
+    var refreshRequest = URLRequest(url: url)
+    refreshRequest.httpMethod = "POST"
+    refreshRequest.addValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
+    refreshRequest.addValue("application/json", forHTTPHeaderField: "Accept")
+    
+    let queryString = params.map({ key, value in
+        let k = key.addingPercentEncoding(withAllowedCharacters: .formUrlEncoded)!
+        let v = value.addingPercentEncoding(withAllowedCharacters: .formUrlEncoded)!
+        return "\(k)=\(v)"
+    }).joined(separator: "&")
+    
+    let formData = queryString.data(using: .utf8)
+    
+    refreshRequest.httpBody = formData
+    
+    return refreshRequest
+}
+
+public class AbstractOAuthFlowClient: SecurityClient {
+    
+    private let tokenManager: OAuthAccessTokenManager
+    
+    let clientId: String
+    let clientSecret: String
+    
+    public var refreshToken: String? {
+         get async {
+            return await tokenManager.accessToken?.refreshToken
+        }
+    }
+
+    init(clientId: String, clientSecret: String, refreshURL: URL?, automaticallyRefreshTokensThisManySecondsEarly autoRefreshInterval: TimeInterval? = nil) {
+        self.clientId = clientId
+        self.clientSecret = clientSecret
+        self.tokenManager = OAuthAccessTokenManager(refreshTokenURL: refreshURL, automaticallyRefreshTokensThisManySecondsEarly: autoRefreshInterval)
+    }
+    
+    func setAccessToken(_ token: OAuthToken) async throws {
+        try await tokenManager.setAccessToken(token)
+    }
+    
+    func setRefreshToken(_ token: String) async throws {
+        try await tokenManager.setRefreshToken(token)
+        try await tokenManager.refreshToken()
```

Comment:
I feel like this is a bit enigmatic... it makes sense inside the following `authenticate` method... can this code be in the `authenticate` method instead?

Perhaps we shouldn't have this method at all and we should instead have an `OAuthExistingSessionClient` that takes a refresh token and continues the session... it's a bit strange that all of our flows support this, when our use-case for this is just resuming a session. Do we need to care what flow was used to create the session? I don't think we do... so then the app doesn't need to try to reuse the right flow client when resuming a session (such as by getting the refresh token out of the keychain).

Reply:
whether or not a refresh token works depends on the refreshUrl supplied by the pre defined flow. It seems to me that the consumer of the API will need to know which flow they were on before supplying a refresh token.